boot/spring_performance.go

```diff
@@ -0,0 +1,265 @@
+		DoTrainingRun:    doTrainingRun,
+		ClasspathString:  classpathString,
+		ReZip:            reZip,
+	}
+}
+
+func (s SpringPerformance) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
+	s.LayerContributor.Logger = s.Logger
+	layer, err := s.LayerContributor.Contribute(layer, func() (libcnb.Layer, error) {
+
+		layer.LaunchEnvironment.Default("BPL_SPRING_AOT_ENABLED", s.AotEnabled)
+
+		if !s.DoTrainingRun {
+			return layer, nil
+		}
+
+		layer.LaunchEnvironment.Default("BPL_JVM_CDS_ENABLED", s.DoTrainingRun)
+
+		// prepare the training run JVM opts
+		var trainingRunArgs []string
+
+		if s.AotEnabled {
+			trainingRunArgs = append(trainingRunArgs, "-Dspring.aot.enabled=true")
+		}
+
+		jarPath := s.AppPath
+
+		if s.ReZip {
+			jarDestDir := os.TempDir() + "/" + fmt.Sprint(time.Now().UnixMilli()) + "/jar-dest"
+			if err := os.MkdirAll(jarDestDir, 0755); err != nil {
+				return layer, fmt.Errorf("error creating temp directory for jar\n%w", err)
+			}
+			tempJarPath := filepath.Join(jarDestDir, "runner.jar")
+			if err := CreateJar(s.AppPath+"/", tempJarPath); err != nil {
+				return layer, fmt.Errorf("error recreating jar\n%w", err)
+			}
+			f, err := os.Open(tempJarPath)
+			if err != nil {
+				return layer, fmt.Errorf("error opening jar\n%w", err)
+			}
+			if err = sherpa.CopyFile(f, filepath.Join(layer.Path, "runner.jar")); err != nil {
+				return layer, fmt.Errorf("error copying jar\n%w", err)
+			}
+
+			jarPath = tempJarPath
+			os.RemoveAll(s.AppPath)
+		}
+
+		if err := s.springBootJarCDSLayoutExtract(jarPath); err != nil {
+			return layer, fmt.Errorf("error extracting Boot jar at %s\n%w", jarPath, err)
+		}
+		startClassValue, _ := s.Manifest.Get("Start-Class")
+
+		err := resetCreationTimeWithTouch(s)
+		if err != nil {
+			return libcnb.Layer{}, fmt.Errorf("error resetting file creation time for CDS\n%w", err)
+		}
+
+		trainingRunArgs = append(trainingRunArgs,
+			"-Dspring.context.exit=onRefresh",
+			"-XX:ArchiveClassesAtExit=application.jsa",
+			"-cp",
+		)
+		trainingRunArgs = append(trainingRunArgs, s.ClasspathString)
+		trainingRunArgs = append(trainingRunArgs, startClassValue)
+
+		var trainingRunEnvVariables []string
+		
+		javaToolOptions := sherpa.GetEnvWithDefault("CDS_TRAINING_JAVA_TOOL_OPTIONS", sherpa.GetEnvWithDefault("JAVA_TOOL_OPTIONS", ""))
+		if javaToolOptions != "" {
+			s.Logger.Bodyf("Training run will use this value as JAVA_TOOL_OPTIONS: %s", javaToolOptions)
+			trainingRunEnvVariables = append(trainingRunEnvVariables, fmt.Sprintf("JAVA_TOOL_OPTIONS=%s", javaToolOptions))
+		}
+
+		// perform the training run, application.dsa, the cache file, will be created
+		if err := s.Executor.Execute(effect.Execution{
+			Command: "java",
+			Env:     trainingRunEnvVariables,
+			Args:    trainingRunArgs,
+			Dir:     s.AppPath,
+			Stdout:  s.Logger.InfoWriter(),
+			Stderr:  s.Logger.InfoWriter(),
+		}); err != nil {
+			return libcnb.Layer{}, fmt.Errorf("error running build\n%w", err)
+		}
```

Comment:
well, we've run a whole lot of tests (at least 6 differentpeople with different JVM, hardware, OSes, etc.) and... that never was a concerned.

The "magic" is under this [Spring Framework 6 innovation](https://github.com/spring-projects/spring-framework/issues/30606): `-Dspring.context.exit=onRefresh` - when the beans are loaded, pouf, the app exits - and there's no way a working app would hung before that.

Reply:
I'm ok with it if you're confident in that it won't happen. A hung build can be a big deal for users though, for example it can be potentially costly since many people pay by the minute for CI/build time, so I just wanted to mention it. Make sure we're thinking through that scenario.